src/keys/drivers/YubiKey.cpp

```diff
@@ -54,161 +132,162 @@ YubiKey* YubiKey::instance()
     return m_instance;
 }
 
-bool YubiKey::init()
+bool YubiKey::isInitialized()
 {
-    m_mutex.lock();
-
-    // previously initialized
-    if (m_yk != nullptr && m_ykds != nullptr) {
-
-        if (yk_get_status(m_yk, m_ykds)) {
-            // Still connected
-            m_mutex.unlock();
-            return true;
-        } else {
-            // Initialized but not connected anymore, re-init
-            deinit();
-        }
-    }
+    return m_initialized;
+}
 
-    if (!yk_init()) {
-        m_mutex.unlock();
-        return false;
+void YubiKey::findValidKeys()
+{
+    m_error.clear();
+    if (!isInitialized()) {
+        return;
     }
 
-    // TODO: handle multiple attached hardware devices
-    m_onlyKey = false;
-    m_yk_void = static_cast<void*>(yk_open_first_key());
-#if YKPERS_VERSION_NUMBER >= 0x011400
-    // New fuction available in yubikey-personalization version >= 1.20.0 that allows
-    // selecting device VID/PID (yk_open_key_vid_pid)
-    if (m_yk == nullptr) {
-        static const int device_pids[] = {0x60fc}; // OnlyKey PID
-        m_yk_void = static_cast<void*>(yk_open_key_vid_pid(0x1d50, device_pids, 1, 0));
-        m_onlyKey = true;
-    }
-#endif
-    if (m_yk == nullptr) {
-        yk_release();
-        m_mutex.unlock();
-        return false;
-    }
+    QtConcurrent::run([this] {
+        if (!m_mutex.tryLock(1000)) {
+            emit detectComplete(false);
+            return;
+        }
 
-    m_ykds_void = static_cast<void*>(ykds_alloc());
-    if (m_ykds == nullptr) {
-        yk_close_key(m_yk);
-        m_yk_void = nullptr;
-        yk_release();
-        m_mutex.unlock();
-        return false;
-    }
+        // Remove all known keys
+        m_foundKeys.clear();
+
+        // Try to detect up to 4 connected hardware keys
+        for (int i = 0, j = 0; i + j < 4;) {
+            bool onlyKey = false;
+            auto* yk_key = openKey(i, j, &onlyKey);
+            Tools::wait(50);
```

Comment:
What are we waiting for here? This looks an awful lot like "no idea what's happening here, let's just hope this is enough to not fail randomly on some persons's machine". Isn't there a better solution?

Reply:
There isn't, removing these waits causes sporadic issues reading from the keys. This one *may* be able to be removed since I excluded yubikey neo from the CR testing.

Reply:
> since I excluded yubikey neo from the CR testing

Will this be of any problem for "Neo" users? Just curious if I need to replace mine beforehand.